cpp/include/cudf/detail/utilities/stream_pool.hpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+
+namespace cudf::detail {
+
+/**
+ * @brief A pool of CUDA stream objects
+ *
+ * Provides efficient access to a collection of asynchronous (i.e. non-default) CUDA stream objects.
+ *
+ * The default implementation uses an underlying `rmm::cuda_stream_pool`. The only other
+ * implementation at present is a debugging version that always returns the stream returned by
+ * `cudf::get_default_stream()`. To use this debugging version, set the environment variable
+ * `LIBCUDF_USE_DEBUG_STREAM_POOL`.
+ *
+ * Access to the global `cuda_stream_pool` is granted via `cudf::detail::global_cuda_stream_pool()`.
+ *
+ * Example usage:
+ * @code{.cpp}
+ * auto stream = cudf::get_default_stream();
+ * auto const num_streams = 2;
+ * // do work on stream
+ * auto streams = cudf::detail::global_cuda_stream_pool().get_streams(num_streams);
```

Comment:
Let's consider a simpler API. What pros/cons do you see here?

```c++
auto stream = cudf::get_default_stream();
auto const num_streams = 2;
auto streams = cudf::detail::fork_streams(stream, num_streams);
// do work on streams[0], streams[1]
cudf::detail::join_streams(streams);
```

Here, the `fork_streams` call would return a vector of streams taken from the global pool -- no need to refer to the global pool directly at all.

Reply:
I also considered (and rejected) making `fork_streams` return an RAII object that automatically joins. That might be too much for readers to expect implicit joining -- and maybe not stream-safe by design, if it encourages bad habits where host objects go out of scope before the join occurs... that's already a case where we expect to require a manual stream sync, but it'd require an explicit join before that sync happens as well.

Reply:
I'm not opposed.  @vuule?

Reply:
Looks neat. Would this be an overload of fork_streams or a replacement? I wonder if we ever need the original, to fork streams that don't come directly from the pool.

Reply:
And should we keep `get_streams()` in the stream pool interface, or just roll that functionality into `fork_streams()`?

Reply:
I vote to replace the existing fork_streams. And to keep the get_streams pool interface, just internally for use by fork_streams.

I tried to think of cases where you’d want a pool without forking and still want to sync with another stream, but I didn’t see anything immediately.

Reply:
@bdice I've pushed a new minimalist interface...is this what you were after, or is this too far?

Reply:
I think the new interface is good; pretty hard to misuse.
I've thought about whether we'd want to keep get and fork separate, but could not find a use case that can't use the current API. So +1 for the simple API from me :)